connectors/slack/template/configure.js

```diff
@@ -1,84 +1,117 @@
-                    ...state.data, 
-                    slack_access_token: response.body.access_token
-                };
-                return Sdk.completeWithSuccess(state, data);
-            }
-            else {
-                throw { 
-                    status: 500, 
-                    message: `Slack authorization failed: ${ctx.query.error_description || ctx.query.error || 'Unknown error'}`, 
-                    state 
-                };
-            }
+      if (ctx.query.code) {
+        const selfUrl = Sdk.getSelfUrl(ctx);
+        let response;
+        try {
+          response = await Superagent.post(`https://slack.com/api/oauth.v2.access`).type('form').send({
+            code: ctx.query.code,
+            client_id: ctx.configuration.slack_client_id,
+            client_secret: ctx.configuration.slack_client_secret,
+            redirect_uri: selfUrl,
+          });
+          if (!response || !response.body || !response.body.access_token) {
+            throw new Error('Unable to obtain access token');
+          }
+        } catch (e) {
+          throw {
+            status: 500,
+            message: `Slack authorization failed: ${e.message}`,
+            state,
+          };
         }
-    }
+
+        let data = {
+          ...state.data,
+          slack_bot_access_token: response.body.access_token,
+          slack_bot_id: response.body.bot_user_id,
+          slack_user_access_token: response.body.authed_user && response.body.authed_user.access_token,
+          slack_user_id: response.body.authed_user && response.body.authed_user.id,
+          slack_app_id: response.body.app_id,
+          slack_team_id: response.body.team && response.body.team.id,
+        };
+
+        // To support Slack's Events API, we use this function as a router that the user registers
+        // with Slack. So we need to be able to route incoming events to all addon instances that are
+        // created for users of this app + team (workspace). The way this works is as follows:
+        //
+        // 1. Here we generate and store a 2-use token associated with the app+team that the user
+        // just authorized. It is then sent back in the config blob along with the registration URL
+        // to call with that token.
+        //
+        // 2. During addon install, the addon template makes the first use of the token to register
+        // the addon URL it just created to receive events
+        //
+        // 3. During addon uninstall, it makes the second call and final token use to deregister
+        //
+        if (ctx.configuration.slack_signing_secret) {
+          const tokenLength = 48;
+          const tokenUses = 2;
+
+          storage.ensureStorage(ctx);
+          await storage.ensureCache();
+          let token = (await crypto.randomBytes(tokenLength)).toString('hex');
+          await storage.put(() => {
+            storage.tokenToTeam[token] = [`${data.slack_app_id}-${data.slack_team_id}`, tokenUses];
```

Comment:
Why don't you store an object with three properties, or an array with three elements instead?

Reply:
Fixed